integration_test/tests/disconnect_tests.rs

```diff
@@ -0,0 +1,172 @@
+                ),
+            );
+
+            // Should be a no-op since the statement is not bound. This will hang if SQLDisconnect does not close the cursor
+            assert_eq!(
+                SqlReturn::SUCCESS,
+                SQLFreeStmt(stmt, FreeStmtOption::SQL_UNBIND as i16),
+                "{}",
+                get_sql_diagnostics(
+                    HandleType::SQL_HANDLE_STMT,
+                    *ptr::addr_of!(stmt).cast::<Handle>()
+                ),
+            );
+
+            disconnect_and_free_dbc_and_env_handles(env, dbc)
+        }
+    }
+
+    #[test]
+    fn sql_disconnect_handles_many_statements_properly() {
+        let env = allocate_env(AttrOdbcVersion::SQL_OV_ODBC3);
+        let dbc = connect_with_conn_string(env, None)
+            .expect("Failed to connect with default connection string");
+        // The following statements will be allocated and interacted with in the following ways:
+        // will be closed
+        let stmt_1 = allocate_statement(dbc).expect("Failed to allocate statement 1");
+        // will be unbound
+        let stmt_2 = allocate_statement(dbc).expect("Failed to allocate statement 2");
+        // will be closed and have its handle deallocated
+        let stmt_3 = allocate_statement(dbc).expect("Failed to allocate statement 3");
+        // will have no operations performed on it
+        let stmt_4 = allocate_statement(dbc).expect("Failed to allocate statement 4");
+        let statements = [stmt_1, stmt_2, stmt_3, stmt_4];
+
+        let query = b"SELECT * FROM integration_test.foo\0".map(|b| b as u16);
+        statements.iter().for_each(|stmt| unsafe {
+            assert_eq!(
+                SqlReturn::SUCCESS,
+                SQLExecDirectW(*stmt, query.as_ptr(), SQL_NTS as i32,),
+                "{}",
+                get_sql_diagnostics(
+                    HandleType::SQL_HANDLE_STMT,
+                    *ptr::addr_of!(stmt).cast::<Handle>()
+                ),
+            );
+        });
+
+        unsafe {
+            // Close statement 1
+            assert_eq!(
+                SqlReturn::SUCCESS,
+                SQLFreeStmt(stmt_1, FreeStmtOption::SQL_CLOSE as i16),
+                "{}",
+                get_sql_diagnostics(
+                    HandleType::SQL_HANDLE_STMT,
+                    *ptr::addr_of!(stmt_1).cast::<Handle>()
+                ),
+            );
+
+            // Unbind statement 2
+            assert_eq!(
+                SqlReturn::SUCCESS,
+                SQLFreeStmt(stmt_2, FreeStmtOption::SQL_UNBIND as i16),
+                "{}",
+                get_sql_diagnostics(
+                    HandleType::SQL_HANDLE_STMT,
+                    *ptr::addr_of!(stmt_2).cast::<Handle>()
+                ),
+            );
+
+            // Close statement 3
+            assert_eq!(
+                SqlReturn::SUCCESS,
+                SQLFreeStmt(stmt_3, FreeStmtOption::SQL_CLOSE as i16),
```

Comment:
This step doesn't look mandatory if you look at the transition table for statements.
Both SQLDisconnect and SQLFreeHandle transition table look the same for statements.

Reply:
I was just putting statements in different states to exercise the iteration and show that we won't crash when going through them.

Reply:
yes, that part is great 👍  
I'm just saying that you don't need to do :
1. SQLFreeStmt(stmt_3)
2. SQLFreeHandle(stmt_3)
You could just call SQLFreeHandle(stmt_3) directly since SQLFreeStmt will get you from S4 to S1, but  SQLFreeHandle gets you to S0 from S4 or S1 just the same.

If you are trying to check that we can call SQLFreeHandle(stmt_3) after SQLFreeStmt(stmt_3) maybe this could be a test in free_and_cancel_tests.rs?